app/grandchallenge/components/tasks.py

```diff
@@ -632,8 +635,16 @@ def provision_job(
     )
     executor = job.get_executor(backend=backend)
 
+    if not job.inputs_complete:
+        logger.warning("Nothing to do: inputs are not complete.")
+        return
+
     if job.status not in [job.PENDING, job.RETRY]:
-        raise RuntimeError("Job is not ready for provisioning")
+        logger.error(
+            "Job is not ready for provisioning, or has already been provisioned",
+            exc_info=True,
+        )
+        return
```

Comment:
Why is this change? This should raise an error here - nothing has been altered in the task, so there is nothing to rollback. If it does need to change the `exc_info=True` is superfluous as there is no unhandled exception here.

Reply:
Hm, I don't think we want to raise an error here. This task can be called multiple times with complete inputs. The first time it's called, the job is locked, inputs are provisioned and the state is updated. The second time it is called the state will no longer be pending (or retry), but we don't want to raise an error in that case. It is expected, we just want to exit and move on. So no need to raise an exception and the logger level should be info. Would you agree?

Reply:
We could check `job.inputs_complete` again here and only raise an error if that is not the case. Then we'd catch the unexpected cases.

Reply:
Ah no that makes no sense, since the code won't reach that point if the inputs are not complete.

Reply:
I still think that we don't need to raise the exception here. Can you explain when that would be necessary?

Reply:
Because the task should not be called unless the job is in the correct state. If the job is not in the correct state and this task has been called it means that something has gone wrong in our application and we must know about it.

Reply:
Since we don't want to use Celery chains and groups anymore, I think the only solution to guarantee that we only call `provision_job` once is to have 1 celery task that does all input creation and validation (rather than individual celery tasks per input). This task would need to run on the large queue. We could theoretically run into timeout issues for jobs with lots of very large image inputs, but I guess this would be very rare.

Reply:
After discussing this synchronously in our meeting today, I did not move to using 1 single task for all inputs, but instead added a new status and the necessary checks in the task that runs before `provision_inputs` to ensure it runs only once.